src/linker/Linker/LinkerAttributesInformation.cs

```diff
@@ -5,49 +5,62 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
+using System.Linq;
 using Mono.Cecil;
 
 namespace Mono.Linker
 {
-	struct LinkerAttributesInformation
+	readonly struct LinkerAttributesInformation
 	{
-		Dictionary<Type, object> _linkerAttributes;
+		readonly Dictionary<Type, List<Attribute>> _linkerAttributes;
 
-		public void InitializeForMethod (MethodDefinition method)
+		public LinkerAttributesInformation (MethodDefinition method)
 		{
-			if (method.HasCustomAttributes)
+			_linkerAttributes = null;
+
+			if (method.HasCustomAttributes) {
 				foreach (var customAttribute in method.CustomAttributes) {
 					var attributeType = customAttribute.AttributeType;
+					Attribute attributeValue = null;
 					if (attributeType.Name == "RequiresUnreferencedCodeAttribute" && attributeType.Namespace == "System.Diagnostics.CodeAnalysis") {
-						AddAttribute (ProcessRequiresUnreferencedCodeAttribute (customAttribute));
+						attributeValue = ProcessRequiresUnreferencedCodeAttribute (customAttribute);
 					}
-				}
-		}
 
-		public bool TryGetAttribute<T> (out T attributeValue) where T : class
-		{
-			attributeValue = null;
+					if (attributeValue != null) {
+						if (_linkerAttributes == null)
+							_linkerAttributes = new Dictionary<Type, List<Attribute>> ();
+
+						Type attributeValueType = attributeValue.GetType ();
+						if (!_linkerAttributes.TryGetValue (attributeValueType, out var attributeList)) {
+							attributeList = new List<Attribute> ();
+							_linkerAttributes.Add (attributeValueType, attributeList);
+						}
 
-			if (_linkerAttributes != null && _linkerAttributes.TryGetValue (typeof (T), out var returnValue)) {
-				attributeValue = (T) returnValue;
-				return true;
+						attributeList.Add (attributeValue);
+					}
+				}
 			}
+		}
 
-			return false;
+		public bool HasAttribute<T> () where T : Attribute
+		{
+			return _linkerAttributes != null && _linkerAttributes.ContainsKey (typeof (T));
```

Comment:
```suggestion
			return _linkerAttributes?.ContainsKey (typeof (T));
```

Reply:
Such expression returns bool? and so I would have to also add == true  - which honestly makes the whole thing less readable.